File: scripts/tokenizer.js
#!/usr/bin/node

"use strict";

const util = require('util');
const strenc = new util.TextEncoder('utf-8');
const {subtle} = require('crypto').webcrypto;

var readline = require('readline');
var reader = readline.createInterface({input: process.stdin, output: process.stdout, terminal: true});
var prompt = (query) => new Promise((resolve) => reader.question(query, resolve));

(async () => {

	const buff2hex = function(buffer) { return [...new Uint8Array(buffer)].map(x => x.toString(16).padStart(2, '0')).join(''); }

	const buff2pin = function(token_buffer)
	{
		var hex_data = buff2hex(token_buffer);
		var hex_word = hex_data.match(/(.{1,6})/g);			
		var hexpin = parseInt(hex_word.shift(), 16);

		for (var i=0; i < hex_word.length; i++) hexpin = (hexpin ^ parseInt(hex_word[i], 16));
		return hexpin;
	}

	const tokenizer = function(auth, host, key)
	{
		const pbkdf2_pref = {name: 'PBKDF2', hash: 'SHA-256', salt: strenc.encode(key), iterations: 300000};
		const hmac_pref = {name: 'HMAC', hash: 'SHA-512'};

		return new Promise(function(resolve)
		{
			subtle.importKey('raw', strenc.encode(auth), 'PBKDF2', false, ['deriveBits'])
			.then(pbkdf2_key => subtle.deriveBits(pbkdf2_pref, pbkdf2_key, 64*8))
			.then(pbkdf2_hash => subtle.importKey('raw', strenc.encode(buff2hex(pbkdf2_hash)), hmac_pref, false, ['sign']))
			.then(hmac_key => subtle.sign('HMAC', hmac_key, strenc.encode(host)))
			.then(hmac_hash => subtle.digest('SHA-1', hmac_hash))
			.then(function(token_sha1)
			{
				var token = btoa(String.fromCharCode(...new Uint8Array(token_sha1)));
				var pin = token.substring(token.length - 4) + buff2pin(token_sha1);

				resolve({token: token, pin: pin});
			});
		})
	}

	const digest = async function(user, login, server, password)
	{
		const key = await subtle.digest('SHA-512', strenc.encode(password));
		const host = await subtle.digest('SHA-512', strenc.encode(login + server));
		const auth = await subtle.digest('SHA-512', strenc.encode(user + buff2hex(host)));

		return {auth: auth, host: host, key: key};
	}

	//--[ User inputs ]-----------------------------------------------

	const user = await prompt("User: ");

	console.log("Password:");

	reader._writeToOutput = function(input){this.output.write('')};
	const password = await prompt("\n");
	reader._writeToOutput = function(input){this.output.write(input)};

	const login = await prompt("\nLogin: ");
	const server = await prompt("Server: ");

	//--[ SHA512 digest + Tokenize with PBKDF2 / HMAC ]---------------

	digest(user, login, server, password)
	.then(hash => tokenizer(buff2hex(hash.auth), buff2hex(hash.host), buff2hex(hash.key)))
	.then(function(results)
	{
		console.log("\nToken: !" + results.token);
		console.log("PIN: !" + results.pin);
	});

	reader.close();
})()

File: scripts/tokenizer.py
#!/usr/bin/python3

import getpass
import hashlib
import base64
import hmac

#--[ User inputs ]------------------------------------------------------------------------------------------

user = input("User: ")
password = getpass.getpass()

login = input("\nLogin: ")
host = input("Server: ")

#--[ SHA512 digest inputs ]---------------------------------------------------------------------------------

hash_password = hashlib.sha512(password.encode('utf-8')).hexdigest()
hash_host = hashlib.sha512(login.encode('utf-8') + host.encode('utf-8')).hexdigest()
hash_user = hashlib.sha512(user.encode('utf-8') + hash_host.encode('utf-8')).hexdigest()

#--[ Tokenize with PBKDF2/HMAC ]----------------------------------------------------------------------------

pbkdf2 = hashlib.pbkdf2_hmac('sha256', hash_user.encode('utf-8'), hash_password.encode('utf-8'), 300000, 64)
hmac512 = hmac.new(pbkdf2.hex().encode('utf-8'), hash_host.encode('utf-8'), hashlib.sha512)

#--[ SHA1 + BASE64 for convenient output ]------------------------------------------------------------------

sha1_token = hashlib.sha1(hmac512.digest())
b64_token = base64.b64encode(sha1_token.digest()).decode('utf-8')

#--[ User output ]------------------------------------------------------------------------------------------

print("\nToken: !" + b64_token)

#--[ PIN alternative ]--------------------------------------------------------------------------------------

hex_word = []
hex_token = sha1_token.hexdigest().upper()

for i in range(0, len(hex_token), 6): hex_word.append(hex_token[i:i+6])

pin_eval = '^0x'.join(str(n) for n in hex_word)
pin_token = eval('0x' + pin_eval)

print("PIN: " + b64_token[-4:] + str(pin_token))
